---
title: "data prep"
author: "Laura Garc√≠a"
date: "2025-07-02"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(tidyr)
```

```{r}
annot <- readRDS("Potra02_genes.rds")
gene_id <- annot$Name %>% 
  as.data.frame()

 gene_id <-  rename(gene_id, id = `.`) 
 gene_id <- gene_id %>% 
    drop_na() %>% separate(id, c("id", NA), "\\.") %>% 
    unique() %>% 
    pull(id)
# saveRDS(gene_id, file = "annotations/gene_id.rds")

# #Filter annot: I will only use gene, mRNA and exons
# annot <- annot[annot$type == "gene"|  
#                  annot$type == "mRNA"|
#                  annot$type == "exon"]


genes <- annot[annot$type == "gene", ]


# Create index dataframe
gene_index <- data.frame(
  geneID = gene_id,
  chromosome = as.character(seqnames(genes)),
  stringsAsFactors = FALSE
)

# Save index
saveRDS(gene_index, "annotations/gene_index.rds")
gene_index <- readRDS("annotations/gene_index.rds")

# gene_index_sel <- gene_index %>% 
#   filter(chromosome == "chr8") %>% 
#   filter(grepl("15", geneID))

genes_sel <- c("Potra2n8c16715",
"Potra2n8c16915",
"Potra2n8c17015",
"Potra2n8c17115",
"Potra2n8c17151",
"Potra2n8c17152",
"Potra2n8c17155",
"Potra2n8c17157",
"Potra2n8c17158",
"Potra2n8c17159",
"Potra2n8c17215",
"Potra2n8c17315",
"Potra2n8c18153",
"Potra2n8c18159",
"Potra2n8c18215")

gene_index_sel <- gene_index[gene_index$geneID %in% genes_sel,]
saveRDS(gene_index_sel, "annotations/gene_index_sel.rds")
gene_index_sel <- readRDS("annotations/gene_index_sel.rds")
# saveRDS(annot, file = "RData/Potra02_genes.rds")

#Split annot by chromosomes
chromosomes <- unique(seqnames(annot))

# Create directory for split files
# dir.create("annotation/by_chromosome", showWarnings = FALSE)

# Split and save by chromosome
lapply(chromosomes, function(chr) {
  chr_annot <- annot[seqnames(annot) == chr, ]
  saveRDS(chr_annot, file.path("annotations", paste0(chr, ".rds")))
})

#Filter only a few genes
chr_annot <- readRDS("annotations/chr8.rds")
root_id <- sub("\\..*$", "", chr_annot$ID)
chr_annot_sel <- chr_annot[root_id %in% gene_index_sel$geneID]

#Or
root_id <- sub("\\..*$", "", annot$ID)
chr_annot_sel <- annot[root_id %in% gene_index_sel$geneID]

#Increase 8000bp downstream and upstream
chr_annot_sel <- chr_annot_sel + 10000

saveRDS(chr_annot_sel, "annotations/chr8_sel.rds")

annot_sel <- readRDS("annotations/chr8_sel.rds")

file_groups <- list(
  rna = c(
    "split_bw/RNA-seq/LD.bw",
    "split_bw/RNA-seq/SDW4.bw",
    "split_bw/RNA-seq/SDW10.bw",
    "split_bw/RNA-seq/CTW6.bw",
    "split_bw/RNA-seq/LD10d.bw"
  ),
  atac = c(
    "split_bw/ATAC/LD.bw",
    "split_bw/ATAC/SDW4.bw",
    "split_bw/ATAC/SDW10.bw",
    "split_bw/ATAC/CTW6.bw",
    "split_bw/ATAC/LD10d.bw"
  ),
  k4 = c(
    "split_bw/H3K4me3/LD.bw",
    "split_bw/H3K4me3/SDW4.bw",
    "split_bw/H3K4me3/SDW10.bw",
    "split_bw/H3K4me3/CTW6.bw",
    "split_bw/H3K4me3/LD10d.bw"
  ),
  k27me = c(
    "split_bw/H3K27me3/LD.bw",
    "split_bw/H3K27me3/SDW4.bw",
    "split_bw/H3K27me3/SDW10.bw",
    "split_bw/H3K27me3/CTW6.bw",
    "split_bw/H3K27me3/LD10d.bw"
  ),
  k27ac = c(
    "split_bw/H3K27ac/LD.bw",
    "split_bw/H3K27ac/SDW4.bw",
    "split_bw/H3K27ac/SDW10.bw",
    "split_bw/H3K27ac/CTW6.bw",
    "split_bw/H3K27ac/LD10d.bw"
  )
)
# # Function to import and save as RDS
# process_files <- function(file_list, group_name) {
#   lapply(seq_along(file_list), function(i) {
#     message("Processing: ", file_list[i])
#     
#     # Import the BigWig file
#     bw_data <- import(file_list[i])
#     
#     # # Convert to Rle-packed format to save space
#     # compressed_data <- as(bw_data, "Rle")
#     
#     # Generate output filename
#     out_name <- paste0("data/", group_name, "_", i, ".rds")
#     
#     # Save with maximum compression
#     saveRDS(bw_data, file = out_name)
#     
#     message("Saved: ", out_name)
#   })
# }
# 
# # Process all files
# mapply(process_files, file_groups, names(file_groups))

# Function to process one BigWig file
process_bw_file <- function(bw_path, group_name) {
  # Extract sample name (LD, SDW4, etc.)
  sample_name <- tools::file_path_sans_ext(basename(bw_path))
  
  # Create group directory (RNA-seq, ATAC, etc.)
  group_dir <- file.path(
    "RData", group_name)
  dir.create(group_dir, showWarnings = FALSE, recursive = TRUE)
  
  # Import BigWig data
  message("Processing: ", bw_path)
  gr <- import(bw_path)
  
  # Split by chromosome and save
  chroms <- unique(seqnames(gr))
  lapply(chroms, function(chr) {
    chr_data <- gr[seqnames(gr) == chr]
    out_file <- file.path(group_dir, paste0(sample_name, "_", chr, ".rds"))
    saveRDS(chr_data, file = out_file)
  })
  
  message("Saved processed data for: ", group_name, "/", sample_name)
}

mapply(function(group_files, group_name) {
  lapply(group_files, process_bw_file, group_name = group_name)
}, file_groups, names(file_groups))

#Filter chr8 files for a few genes                         
rds_files <- list.files(
  "RData", pattern = "\\.rds$", recursive = TRUE, full.names = TRUE
)

lapply(rds_files, function(f) {
  gr          <- readRDS(f)
  gr_sel      <- subsetByOverlaps(gr, annot_sel, ignore.strand = TRUE)
  out_path    <- sub("\\.rds$", "_sel.rds", f)   # e.g. LD_chr8_sel.rds
  saveRDS(gr_sel, out_path)
})

```

